pkg/verification/verification_service.go

```diff
@@ -87,3 +95,60 @@ func generateVerificationCode() (string, error) {
 
 	return string(buf), nil
 }
+
+// VerifyCode validates the user's phone verification code.  It updates the specified UserSignup value, so even
+// if an error is returned by this function the caller should still process changes to it
+func (s *ServiceImpl) VerifyCode(ctx *gin.Context, signup *v1alpha1.UserSignup, code string) error {
+
+	now := time.Now()
+
+	// If 24 hours has passed since the verification timestamp, then reset the timestamp and verification attempts
+	ts, err := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey])
+	if err != nil || (err == nil && now.After(ts.Add(24*time.Hour))) {
+		// Set a new timestamp
+		signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey] = now.Format(TimestampLayout)
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(0)
+	}
+
+	attemptsMade, err := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
+	if err != nil {
+		// We shouldn't get an error here, but if we do, we should probably set verification attempts to max allowed
+		// so that we at least now have a valid value
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(s.config.GetVerificationAttemptsAllowed())
+		return err
+	}
+
+	// If the user has made more attempts than is allowed per day, return an error
+	if attemptsMade >= s.config.GetVerificationAttemptsAllowed() {
+		return errors2.NewTooManyRequestsError("too many verification attempts")
+	}
+
+	exp, err := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
+	if err != nil {
+		// If the verification expiry timestamp is corrupt or missing, then return an error
+		return errors2.NewInternalError(err)
+	}
+
+	if now.After(exp) {
+		// If it is now past the expiry timestamp for the verification code, return a 400 Bad Request error
+		return errors2.NewBadRequest("verification code expired")
+	}
+
+	// If the code matches then set VerificationRequired to false, reset other verification annotations
+	if code == signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
```

Comment:
Should we make sure that "code" is not an empty string here?

Reply:
i'm wondering if code can even be empty - since it's a path parameter you won't hit the endpoint without it being present, i don't think..

Reply:
Yeah but this is the verification service (not the controller). So, you can still pass an empty "code" to his function. I suggest returning an invalid argument error in this case.

Reply:
In any case I've added an empty check to the controller.